keps/sig-storage/3751-volume-provisioned-io/README.md

```diff
@@ -0,0 +1,473 @@
+
+## Proposal
+
+### Kubernetes API
+Add two well-known ResourceName keys to the existing ResourceList type used in PersistentVolumeClaim.Spec.Resources.{Limits,Requests} and PersistentVolume.Spec.Capacity:
+* ProvisionedIOPS
+* ProvisionedThroughput
+These values can be changed on a PersistentVolumeClaim, and will be reconciled against the storage provider in the same way as storage capacity resizing. The status of a PVC will reflect the currently provisioned IO as reported by the storage provider (in the same way that the status states the current size of the volume during a resize operation).
+
+```
+// Example:
+kind: PersistentVolumeClaim
+...
+spec:
+  ...
+  resources:
+requests:
+  storage: 4Gi
+  throughput: "1.2M"
+  iops: "160000"
+// Vendors may add additional resources using Device Plugin syntax
+// (azure.io/tunable: “100”)
+```
+
+Add PVC Status field; change ResizeStatus alpha field to AllocatedResourceStatus map.
+
+### CSI API
+The CSI create request will be extended to add provisioned IO parameters. A new ControllerUpdateIOProvisioning RPC will be added.
+
+```
+// ControllerServer is the server API for Controller service.
+type ControllerServer interface {
+        ...
+        ControllerUpdateIOProvisioning(context.Context, *ControllerUpdateIOProvisioningRequest) (*ControllerUpdateIOProvisioningResponse, error)
+        ...
+}
+
+type ControllerUpdateIOProvisioningRequest struct {
+        // The ID of the volume to expand. This field is REQUIRED.
+        VolumeId string
+        Iops string 
+        Throughput string
+}
+type ControllerUpdateIOProvisioningResponse struct {
+        Iops int64
+        Throughput int64
+}
+```
+
+
+### Quota
+Resize quota works by using Kubernetes resource quotas, which makes sure the cluster administrators and operators can restrict consumption and creation of cluster resources (such as CPU time, memory, and persistent storage) within a specified namespace. The LimitRange admission controller tracks usage to ensure the storage resource does not exceed resource minimum, maximum and ratio defined in any LimitRange present in the namespace.
+
+```
+apiVersion: v1
+kind: ResourceQuota
+metadata:
+  name: quota
+spec:
+  hard:
+    limits.iops: "200000"    // max for all containers in the namespace
+    limits.throughput: 200
+    requests.iops: "2000"    // max per request 
+    requests.throughput: 20
+    pods: "10"    // max number of pods allowed in the namespace
+```
+
+```
+apiVersion: v1
+kind: LimitRange
+metadata:
+  name: sample
+spec:
+  limits:
+  - max:
+      iops: 1500
+      throughput: 10
+    min:
+      iops: 10
+      throughput: 1
+    default:
+      iops: 100
+      throughput: 5
+    defaultRequest:
+      iops: 100
+      throughput: 20
+    type: Container   
+```
+
+QoS should have similar issues to resize (except that we have to solve reducing IO).
```

Comment:
There should be some description about how current QoS setting in StorageClass parameters will affect namespace quota. Will it get copied (how?) from SCs to PVCs? Will this KEP disallow mixing QoS in StorageClass and in PVCs?

Reply:
Reading the release note - [admission_control_resource_quota.md](https://github.com/kubernetes/kubernetes/blob/release-1.2/docs/design/admission_control_resource_quota.md), it is not so clear to me what is the existing behavior. But I think for the QoS feature, it is risky to trigger an update to force the params within the ResourceQuota etc. Maybe we should design it in a way that does not reduce and only increase Iops/throughput in this situation.